From b66f7203725b4fbbb5fee63601d1c90495adf921 Mon Sep 17 00:00:00 2001
From: Andrew Jeffery <andrew@aj.id.au>
Date: Wed, 12 Feb 2020 12:34:25 +1030
Subject: [PATCH] p10_do_fw_hb_istep: Fix ICE on modern g++
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Cryptically, modern g++, e.g. g++ (Ubuntu 9.2.1-9ubuntu2) 9.2.1
20191008, spit out the following error when building the EKB:

  CXX      ekb/chips/p10/procedures/hwp/istep/libekb_la-p10_do_fw_hb_istep.lo
In file included from ./ekb/hwpf/fapi2/include/ffdc.H:32,
                 from ./ekb/hwpf/fapi2/include/return_code.H:31,
                 from ./ekb/hwpf/fapi2/include/fapi2.H:28,
                 from ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.H:33,
                 from ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C:27:
ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C: In function ‘fapi2::ReturnCode p10_do_fw_hb_istep(const fapi2::Target<fapi2::TARGET_TYPE_PROC_CHIP>&, uint8_t, uint8_t, uint64_t, uint64_t, uint64_t)’:
./hwpf/fapi2/include/plat/plat_trace.H:45:27: internal compiler error: in subspan, at input.h:68
   45 |     printf(_fmt_, ##_args_);                                            \
      |                           ^
./hwpf/fapi2/include/plat/plat_trace.H:54:36: note: in expansion of macro ‘FAPI_TRACE’
   54 | #define FAPI_ERR(_fmt_, _args_...) FAPI_TRACE("err", _fmt_, ##_args_)
      |                                    ^~~~~~~~~~
./hwpf/fapi2/include/plat/plat_error_scope.H:53:9: note: in expansion of macro ‘FAPI_ERR’
   53 |         FAPI_ERR(__VA_ARGS__);                                  \
      |         ^~~~~~~~
./ekb/hwpf/fapi2/include/fapi2_error_scope.H:75:5: note: in expansion of macro ‘PLAT_FAPI_ASSERT’
   75 |     PLAT_FAPI_ASSERT( __conditional__, __ffdc__, __VA_ARGS__ )
      |     ^~~~~~~~~~~~~~~~
ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C:125: note: in expansion of macro ‘FAPI_ASSERT’
  125 |     FAPI_ASSERT((elapsedTimeMs <= i_retry_limit_ms),
      |
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-9/README.Bugs> for instructions.

Interestingly, pre-processing the source then compiling it gives a different
problem:

ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C: In function ‘fapi2::ReturnCode p10_do_fw_hb_istep(const fapi2::Target<fapi2::TARGET_TYPE_PROC_CHIP>&, uint8_t, uint8_t, uint64_t, uint64_t, uint64_t)’:
ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C:125: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 2 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
  125 |     FAPI_ASSERT((elapsedTimeMs <= i_retry_limit_ms),
      |
ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C:199: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 4 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
  199 |     FAPI_ASSERT((elapsedTimeMs <= i_retry_limit_ms),
      |

It turns out fixing the format error also resolves the ICE.

Signed-off-by: Andrew Jeffery <andrew@aj.id.au>
---
 ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C b/ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C
index ac6941de12e3..7a700e173a3b 100644
--- a/ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C
+++ b/ekb/chips/p10/procedures/hwp/istep/p10_do_fw_hb_istep.C
@@ -27,6 +27,8 @@
 #include "p10_do_fw_hb_istep.H"
 #include "p10_scom_proc.H"
 
+#include <inttypes.h>
+
 using namespace scomt::proc;
 
 typedef fapi2::buffer<uint32_t> mbox4_buffer_t;
@@ -126,7 +128,7 @@ fapi2::ReturnCode p10_do_fw_hb_istep(
                 fapi2::DO_FW_HB_ISTEP_NOT_READY()
                 .set_TARGET(i_target)
                 .set_KEY(key),
-                "Ready bit for istep mode wasn't set in %llu ms",
+                "Ready bit for istep mode wasn't set in %" PRIu64 " ms",
                 i_retry_limit_ms);
 
     // Set the go bit for the key.
@@ -206,7 +208,7 @@ fapi2::ReturnCode p10_do_fw_hb_istep(
                 .set_MAX_RETRY_TIME_MS(i_retry_limit_ms)
                 .set_DELAY_MS(i_delay_ms)
                 .set_SIM_CYCLES_DELAY_MS(i_delay_simCycles),
-                "Requested istep %d.%d failed to complete in %llu ms",
+                "Requested istep %d.%d failed to complete in %" PRIu64 " ms",
                 i_istepMajor,
                 i_istepMinor,
                 i_retry_limit_ms);
